taichi/runtime/llvm/runtime.cpp

```diff
@@ -809,13 +809,10 @@ Ptr LLVMRuntime::allocate_from_buffer(std::size_t size, std::size_t alignment) {
   Ptr ret = nullptr;
   bool success = false;
   locked_task(&allocator_lock, [&] {
-    auto alignment_bytes =
-        alignment - 1 -
-        ((std::size_t)preallocated_head + alignment - 1) % alignment;
-    size += alignment_bytes;
-    if (preallocated_head + size <= preallocated_tail) {
-      ret = preallocated_head + alignment_bytes;
-      preallocated_head += size;
+    auto size_aligned = taichi::iroundup(size, alignment);
+    if (preallocated_head + size_aligned <= preallocated_tail) {
+      ret = preallocated_head;
```

Comment:
Looks like we've changed `ret` here, does it matter?

Reply:
we do, that's how it's simplified. The invariant does not change, which is `ret` should be always aligned. Now we can just update `preallocated_head` instead of `preallocated_head`+`alignment_bytes`.